src/Education/XamlGenerator/IRuleHelpXamlBuilder.cs

```diff
@@ -0,0 +1,472 @@
+
+                        case XmlNodeType.Whitespace:
+                            writer.WriteString(reader.Value);
+                            break;
+
+                        case XmlNodeType.EndElement:
+                            XamlOutputElementInfo xamlOutputElement;
+                            do
+                            {
+                                xamlOutputElement = outputXamlElementStack.Pop();
+                                writer.WriteEndElement();
+                            } while (xamlOutputElement.HtmlElementName != reader.Name);
+
+                            break;
+                        default:
+                            var message = string.Format(Resources.XamlBuilder_UnexpectedNodeError, reader.NodeType, reader.Name, reader.Value);
+                            throw new InvalidDataException(message);
+                    }
+                }
+
+                // We've processed all of the html elements.
+                // Now, the stack should only contain the root FlowDocument element, plus an extra
+                // block element if the first tag we processed was an Inline.
+                Debug.Assert(outputXamlElementStack.Count <= 2, "Expecting at most 2 unclosed elements in the stack");
+                Debug.Assert((outputXamlElementStack.Count == 1 && outputXamlElementStack.Peek().HtmlElementName == "html root")
+                    || (outputXamlElementStack.Count == 2 &&
+                        outputXamlElementStack.ToArray()[0].HtmlElementName == null &&
+                        outputXamlElementStack.ToArray()[1].HtmlElementName == "html root"),
+                        "Unexpected items in final stack");
+
+                while (outputXamlElementStack.Count > 0)
+                {
+                    outputXamlElementStack.Pop();
+                    writer.WriteEndElement();
+                }
+            }
+            finally
+            {
+                reader.Close();
+                writer.Close();
+            }
+
+            return sb.ToString();
+        }
+
+
+        private void WriteDocumentHeader(IRuleHelp ruleHelp)
+        {
+            writer.WriteStartElement("FlowDocument", XamlNamespace);
+            writer.WriteAttributeString("xml", "space", null, "preserve");
+            writer.WriteAttributeString("xmlns", "http://schemas.microsoft.com/winfx/2006/xaml/presentation");
+            writer.WriteElementString("Paragraph", $"Rule: {ruleHelp.RuleKey}");
+            writer.WriteElementString("Paragraph", $"Language: {ruleHelp.Language.Name}");
+
+            outputXamlElementStack.Push(new XamlOutputElementInfo("html root", false));
+        }
+
+        private void PushOutputElementInfo(string htmlElementName, bool supportsInlines)
+        {
+            outputXamlElementStack.Push(new XamlOutputElementInfo(htmlElementName, supportsInlines));
+        }
+
+        private void ProcessElement()
+        {
+            switch (reader.Name)
```

Comment:
I might be able to replace most of this code with a simple table lookup that that maps HTML elements to the corresponding XAML element(s).

For now, it's easier to leave it as it is in case I need to add any special logic to handle styling/theming.